### PR TITLE
Minor refactoring of testframe.c testing framework

### PR DESCRIPTION
**Pull request by jhendersonHDF**
Added setup and cleanup callback parameters to TestInit() to perform setup and cleanup tasks once for whole test program

Removed TestCleanup() function since its functionality is covered by PerformTests()

Added check of the HDF5_NOCLEANUP environment variable in GetTestCleanup()

### CODE REVIEW

**Comment by jhendersonHDF** [comment]
A few minor follow-on changes from comments in https://github.com/HDFGroup/hdf5/pull/4891

File: test/testframe.c [hunk 1]
```diff
@@ -387,24 +408,23 @@ TestParseCmdLine(int argc, char *argv[])
...
+herr_t
 PerformTests(void)
 {
     for (unsigned Loop = 0; Loop < TestCount; Loop++) {
         int old_num_errs = TestNumErrs_g;
 
         if (TestArray[Loop].TestSkipFlag) {
-            if (TestFrameworkProcessID_g == 0)
-                MESSAGE(2, ("Skipping -- %s (%s) \n", TestArray[Loop].Description, TestArray[Loop].Name));
+            MESSAGE(2, ("Skipping -- %s (%s) \n", TestArray[Loop].Description, TestArray[Loop].Name));
```
Comment (jhendersonHDF) [inline test/testframe.c:418]: The MESSAGE macro includes the check for `TestFrameworkProcessID_g` of 0

File: test/testframe.c [hunk 2]
```diff
@@ -497,31 +515,26 @@ TestSummary(FILE *stream)
     fprintf(stream, "\n\n");
 }
 
-/*
- * Perform test cleanup
- */
-void
-TestCleanup(void)
```
Comment (jhendersonHDF) [inline test/testframe.c:504]: `TestCleanup()` was only used in a few places and is covered entirely by `PerformTests()` now

File: test/testframe.h [hunk 3]
```diff
@@ -180,15 +197,16 @@ extern "C" {
  *
  */
 H5TEST_DLL herr_t TestInit(const char *ProgName, void (*TestPrivateUsage)(FILE *stream),
-                           int (*TestPrivateParser)(int argc, char *argv[]), int TestProcessID);
+                           int (*TestPrivateParser)(int argc, char *argv[]), herr_t (*TestSetupFunc)(void),
```
Comment (jhendersonHDF) [inline test/testframe.h:200]: Added a run-once setup and cleanup callback for the whole test program. The setup callback is called as part of TestInit() to initialize state as soon as possible and the cleanup callback is called by TestShutdown() to keep the state around as long as possible (for test post-processing or similar).

**Review by byrnHDF — APPROVED** [review]

**Review by mattjala — APPROVED** [review]

**Review by bmribler — APPROVED** [review]

**Review by qkoziol — APPROVED** [review]

**Pull request merged**